kaiax/gov/contractgov/impl/getter.go

```diff
@@ -0,0 +1,89 @@
+package impl
+
+import (
+	"math/big"
+
+	"github.com/kaiachain/kaia/accounts/abi/bind/backends"
+	"github.com/kaiachain/kaia/common"
+	govcontract "github.com/kaiachain/kaia/contracts/contracts/system_contracts/gov"
+	"github.com/kaiachain/kaia/kaiax/gov"
+)
+
+func (c *contractGovModule) EffectiveParamSet(blockNum uint64) (gov.ParamSet, error) {
+	m, err := c.contractGetAllParamsAt(blockNum)
+	if err != nil {
+		return gov.ParamSet{}, err
+	}
+
+	ret := gov.ParamSet{}
+	for k, v := range m {
+		err = ret.Set(k, v)
+		if err != nil {
+			return gov.ParamSet{}, err
+		}
+	}
+	return ret, nil
+}
+
+func (c *contractGovModule) EffectiveParamsPartial(blockNum uint64) (map[gov.ParamEnum]interface{}, error) {
+	return c.contractGetAllParamsAt(blockNum)
+}
+
+func (c *contractGovModule) contractGetAllParamsAt(blockNum uint64) (map[gov.ParamEnum]interface{}, error) {
+	chain := c.Chain
+	if chain == nil {
+		return nil, ErrNotReady
+	}
+
+	config := c.ChainConfig
+	if !config.IsKoreForkEnabled(new(big.Int).SetUint64(blockNum)) {
+		return nil, ErrNotReady
+	}
+
+	addr, err := c.contractAddrAt(blockNum)
+	if err != nil {
+		return nil, err
+	}
+	if common.EmptyAddress(addr) {
+		logger.Trace("ContractEngine disabled: GovParamContract address not set")
+		return nil, nil
+	}
+
+	caller := backends.NewBlockchainContractBackend(chain, nil, nil)
+	contract, _ := govcontract.NewGovParamCaller(addr, caller)
+
+	names, values, err := contract.GetAllParamsAt(nil, new(big.Int).SetUint64(blockNum))
+	if err != nil {
+		logger.Warn("ContractEngine disabled: getAllParams call failed", "err", err)
+		return nil, nil
+	}
+
+	if len(names) != len(values) {
+		logger.Warn("ContractEngine disabled: getAllParams result invalid", "len(names)", len(names), "len(values)", len(values))
+		return nil, nil
+	}
+
+	ret := make(map[gov.ParamEnum]interface{})
+	for i := 0; i < len(names); i++ {
+		param, err := gov.GetParamByName(names[i])
+		if err != nil {
+			return nil, err
+		}
+		cv, err := param.Canonicalizer(values[i])
+		if err != nil {
+			return nil, err
+		}
```

Comment:
1. Absent of `FormatCheck`.
2. Uhm.. even if the `FormatCheck` is added here, the contract storage was already modified without a semantic check (`setParamIn()` does not filter the correctness of value)

Can you share the expected usage scenario?

Reply:
As we cannot perform format check in the contract, the core decides whether the parameters in GovParam is valid or not. If GovParam contains any invalid value, the core ignores GovParam altogether.

Reply:
Actually, I think I'll add a new type similar to `NewVoteData` for contractgov so we don't miss canonicalization and format check.